common/src/primitives/merkle/proof/single/mod.rs

```diff
@@ -0,0 +1,146 @@
+        let leaf_count = tree.leaf_count().get();
+        if leaf_index > leaf_count {
+            return Err(MerkleTreeProofExtractionError::LeafIndexOutOfRange(
+                leaf_index, leaf_count,
+            ));
+        }
+
+        let leaf = tree.node_from_bottom(0, leaf_index).ok_or(
+            MerkleTreeProofExtractionError::AccessError(
+                crate::primitives::merkle::MerkleTreeAccessError::AbsIndexOutOfRange(
+                    leaf_index,
+                    tree.total_node_count().get(),
+                ),
+            ),
+        )?;
+
+        let mut proof = vec![];
+
+        // once we reach root we stop
+        for node in leaf.into_iter_parents() {
+            if node.is_root() {
+                break;
+            }
+            // We push siblings of parents because they're what we need to calculate the root, upwards.
+            let err_msg = "In this loop, this cannot be root, so sibling must exist";
+            let sibling = node.sibling().expect(err_msg);
+            proof.push(sibling);
+        }
+
+        let result = Self {
+            leaf,
+            branch: proof,
+        };
+
+        Ok(result)
+    }
+
+    pub fn into_values(self) -> SingleProofHashes {
+        let proof = self.branch.into_iter().map(|node| *node.hash()).collect::<Vec<_>>();
+        let leaf_abs_index = self.leaf.into_position().position().1 as u32;
+        SingleProofHashes {
+            leaf_index_in_level: leaf_abs_index,
+            branch: proof,
+        }
+    }
+
+    pub fn into_nodes(self) -> Vec<Node<'a>> {
+        self.branch
+    }
+
+    pub fn branch(&self) -> &[Node<'a>] {
+        &self.branch
+    }
+
+    pub fn leaf(&self) -> Node<'a> {
+        self.leaf
+    }
+}
+
+/// Same as `SingleProofNodes`, but has only hashes and leaf index in the lowest level.
+/// This is the minimum information required to prove that the given leaf can produce the root's hash.
+/// This struct is supposed to be serialized, unlike `SingleProofNodes`.
+#[must_use]
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct SingleProofHashes {
+    leaf_index_in_level: u32,
+    branch: Vec<H256>,
+}
+
+impl SingleProofHashes {
+    pub fn into_hashes(self) -> Vec<H256> {
+        self.branch
+    }
+
+    pub fn branch(&self) -> &[H256] {
+        &self.branch
+    }
+
+    pub fn leaf_index_in_level(&self) -> u32 {
+        self.leaf_index_in_level
+    }
+
+    /// Verifies that the given leaf can produce the root's hash.
+    /// Returns None if the proof is empty (the tree has only one node).
+    /// This choice, to return None, is a security measure to prevent a malicious user from
+    /// circumventing verification by providing a proof of a single node.
+    pub fn verify(&self, leaf: H256, root: H256) -> Option<bool> {
+        // in case it's a single-node tree, we don't need to verify or hash anything
+        if self.branch.is_empty() {
+            return None;
+        }
```

Comment:
I don't quite understand the reasoning for the `None` value here. It seems to me single-node tree is a legitimate use case and in that case, we just check `leaf == root`?

Reply:
It's a precaution that I find necessary because the fear is that this gives way too much control possibly with just user input. The problem can happen when the user can provide any input they want for the leaf and prove that "it's equal to the root". This is a trivial case that can lead to vulnerabilities. So, the developer using this function can decide how to handle that case. In the case of taproot, this might be OK because the preimage has to be calculated. But in other cases that I can't foresee, it has to be clear that the test `leaf == root` is a trivial one and possibly fully controlled by user input.

Reply:
Ok, I think I got it. If I understand that correctly, you want to ensure that when `Some(_)` is returned, the prover has demonstrated the knowledge of some hash preimages, so something non-trivial is going on. Something like that?

Reply:
Yep. Exactly.

On the otherhand, the dev can just do `unwrap_or(true)` and it'll just work as if the Option isn't there. The Option will trigger reading the docs of the function to understand this problem (I'll add some more explanation).

Reply:
> On the otherhand, the dev can just do `unwrap_or(true)` and it'll just work as if the Option isn't there.

I think this is one of the shortcomings of the current approach. The devs will be tempted to write something like `.unwrap_or(true)` when it should be `.unwrap_or(leaf == root)`. Maybe just return `Some(false)` if the branch is empty and `leaf != root`?

Reply:
But then what will `None` represent?

Reply:
> But then what will `None` represent?

`Some(true)` = verification passed, `Some(false)` = verification failed, `None` = verification passed trivially. It may be more appropriate to define a new type for this, with arms like `Passed`, `Failed`, `PassedTrivially`. The caller would than have to handle all three cases...

Reply:
Done.